### ISSUE TITLE
Mini-batch size for training

### ISSUE THREAD

**Issue by burcer** [facebookresearch/fairseq-py#84]
The code takes as argument max tokens per mini-batch, but the paper specifies the batchsize to be 64 (smaller if max tokens per mini-batch is exceeded when minibatch size is 64). Is this because one can achieve similar accuracy to what is reported in the paper just by specifying max tokens per mini-batch without worrying about mini-batch size? 
My another question is, what is the maximum mini-batch size (or largest 'max tokens per minibatch') one can use without degrading model quality? For instance, have you tried with say 12000 max tokens per mini-batch (assuming amount of memory is sufficient)?

**Comment by myleott** [comment]
We switched to defining minibatch sizes in terms of the number of tokens (instead of number of sentences) in fairseq-py, since it's more memory efficient and reduces the variance of the gradients.

For single-node 8 GPU training we typically use as many tokens as will fit in GPU memory. For WMT'14 en2de this means 4000 tokens/GPU * 8 GPUs = 32k tokens/batch. We've been exploring distributed training as well (up to 128 GPUs, 512k tokens/batch) and in that case we have to do some warmup of the learning rate to preserve model quality.

**Comment by burcer** [comment]
Thank you for the response!

**Comment by burcer** [comment]
Reopening this because didn't think of asking this earlier: Would you be able to provide the description of learning rate warmup you use for larger tokens/minibatch that you mention above? Sharing the code that you use for that would even be better, of course:)

**Comment by myleott** [comment]
We're still figuring out what learning rate schedules work best, so nothing quite ready-to-share yet, but most of what we're trying is inspired by these works:
- https://arxiv.org/abs/1706.02677
- https://arxiv.org/abs/1711.00489

We will share the code soon :) Basically we've migrated fully to `torch.distributed` and have just finished a large-ish code cleanup. Stay tuned.

**Comment by myleott** [comment]
torch.distributed changes are landed in master.

**Comment by msadat3** [comment]
@myleott what are the differences between --batch-size and --max-sentences?
